x-pack/plugin/indices-write-load-tracker/src/main/java/org/elasticsearch/xpack/indiceswriteloadtracker/IndicesWriteLoadStatsCollector.java

```diff
@@ -0,0 +1,250 @@
+            long totalRefreshTimeInNanos = indexShard.getTotalRefreshTimeInNanos();
+
+            long sampleRelativeTimeInNanos = relativeTimeInNanosSupplier.getAsLong();
+            long samplingTimeInNanos = sampleRelativeTimeInNanos - lastSampleRelativeTimeInNanos;
+            // Even though we're using a MONOTONIC clock (at least System.nanoTime() relies on clock_gettime with CLOCK_MONOTONIC in linux)
+            // it's possible that the clock do not have enough granularity, in that case we bail out early just to be cautious.
+            if (samplingTimeInNanos <= 0) {
+                return;
+            }
+            lastSampleRelativeTimeInNanos = sampleRelativeTimeInNanos;
+
+            long indexingTimeDeltaInNanos = totalIndexingTimeInNanosSample - lastTotalIndexingTimeSample;
+            lastTotalIndexingTimeSample = totalIndexingTimeInNanosSample;
+
+            long mergeTimeDeltaInNanos = totalMergeTimeInNanosSample - lastTotalMergeTimeSample;
+            lastTotalMergeTimeSample = totalMergeTimeInNanosSample;
+
+            long refreshTimeDeltaInNanos = totalRefreshTimeInNanos - lastTotalRefreshTimeSample;
+            lastTotalRefreshTimeSample = totalRefreshTimeInNanos;
+
+            double indexingCPUs = indexingTimeDeltaInNanos / (double) samplingTimeInNanos;
+            double mergeCPUs = mergeTimeDeltaInNanos / (double) samplingTimeInNanos;
+            double refreshCPUs = refreshTimeDeltaInNanos / (double) samplingTimeInNanos;
+
+            synchronized (histogramsLock) {
+                indexingTimeHistogram.recordValue(indexingCPUs);
+                mergeTimeHistogram.recordValue(mergeCPUs);
+                refreshTimeHistogram.recordValue(refreshCPUs);
+            }
+        }
+
+        ShardWriteLoadHistogramSnapshot getWriteLoadHistogramSnapshotAndReset() {
+            synchronized (histogramsLock) {
+                final var indexLoadSummary = new ShardWriteLoadHistogramSnapshot(
+                    getParentDataStreamName(),
+                    indexShard.shardId(),
+                    primary,
+                    new WriteLoadHistogramSnapshot(
+                        System.currentTimeMillis(),
+                        HistogramSnapshot.takeSnapshot(indexingTimeHistogram),
+                        HistogramSnapshot.takeSnapshot(mergeTimeHistogram),
+                        HistogramSnapshot.takeSnapshot(refreshTimeHistogram)
+                    )
+                );
+                indexingTimeHistogram.reset();
+                mergeTimeHistogram.reset();
+                refreshTimeHistogram.reset();
+                return indexLoadSummary;
+            }
+        }
+
+        private String getParentDataStreamName() {
+            String parentDataStream = parentDataStreamName;
+            if (parentDataStream == null) {
+                return (parentDataStreamName = parentDataStreamNameSupplier.get());
```

Comment:
Are we handling data stream modifications or lazy init here? If the former, this is only half doing it then.

Reply:
I think the same as in https://github.com/elastic/elasticsearch/pull/89859#discussion_r968686619  applies here?